Draft terminology related to API behavior definitions (#20959)
zephyrproject-rtos/zephyr: doc/api18970.md

# Draft material to review in support of [#18970][]

[//]: # (cmark-gfm doc/api18970.md > /tmp/api18970.html)

[#18970]: https://github.com/zephyrproject-rtos/zephyr/issues/18970

## Background

See also the Zephyr [glossary][].

[glossary]: https://docs.zephyrproject.org/latest/glossary.html

> ### Thread Terminology

Reviewer: I'm guessing this is just background context that isn't going into the edits to the glossary in the final PR, right?

In other words, I expect the ultimate outcome of this work to be a set of new entries in the glossary, which cross-reference appropriately into the existing kernel docs which define these terms.

Author: This is simply the first step towards the complete requirements of #18970: not just the terminology, but the examples and policies that use that terminology.  Relates to your question about "queued" below: we'll need that term when we discuss the current/desired semantics of an `_async` suffix, perhaps to justify a `_queued` suffix for an API that is asynchronous and non-yielding.

Reviewer: Marking this resolved for now, we can discuss this point further once it's a real docs PR.

historically mean a thread that is *Waiting* for a timeout event.

Zephyr defines [two classes of
threads](https://docs.zephyrproject.org/latest/reference/kernel/threads/index.html#thread-priorities):

* **cooperative** threads remain the current thread until they are made
  unready.
* **preemptible** threads remain the current thread until either (a) a
  cooperative thread becomes *Ready*, or (b) a higher-priority
  preemptible thread becomes *Ready*.

For API behavior we are primarily interested in whether invoking a
function can cause a cooperative thread to yield.

### Context Terminology

**TODO** fill this out

This section is intended to describe the privilege and processor context
variations in which a thread can run.  It should provide or reference
definitions of terms like this:

* Kernel [initialization
  level](https://docs.zephyrproject.org/latest/reference/drivers/index.html#initialization-levels)
* User-space versus kernel (system call)
* Normal ("thread"?) versus interrupt (invoked from an Interrupt Service
  Routine)

For API behavior we are interested in whether a particular function
**may**, **must**, or **must not** be invoked from a specific context.

> ## Definitions

Reviewer: In particular, that the terms in this section will be added to the glossary

Author: Perhaps, perhaps not.  Note that the thread terminology is an an area of documentation about threads; most of this probably belongs in an area of documentation about API specification and behavior.  To be mooted once we have the terminology down.

Reviewer: Ditto here, will mark resolved as out of scope for this draft PR, final resolution TBD

### yield (thread)

A thread yields when it invokes an operation that transitions it to
*Suspended* or *Waiting*.

### pre-empted (thread)

A thread is pre-empted when it transitions from *Running* to *Ready*
without having yielded.  (Only preemptible threads can be pre-empted.)

### yielding (function)

> A function is yielding when it invokes (directly or indirectly) an

Reviewer: ```suggestion
A function is yielding when it may invoke (directly or indirectly) an
```

The "may" addition is important to cover cases like the hypothetical `clock_control_request()`, which might yield until the clock is ready, and might not yield if the clock is already ready.

Author: when speaking of a function generically I meant all paths it might take, so "may" was implicit but your wording is better.

Author: Reworded (significantly).

operation that causes the current thread to yield.

A function is non-yielding when it is not yielding.

### thread-safe

A function is thread-safe if its behavior is correct when invoked from
multiple threads at the same time.

#### Commentary

> Most public API will satisfy this condition; some private API may not.

Reviewer: What condition? Thread safety? I would assume most public API is not thread safe. Interrupt safe is another matter. In general I would expect device access APIs don't automatically lock device data, since that would add unnecessary overhead to single-threaded use. It should generally be the caller's responsibility to ensure exclusive access to such data.

Author: I'm not sure I agree, given the existence of pre-emptible threads, and the possibility (per #20919) that cooperative threads may get booted out in the middle of whatever they're doing too.

Author: Clarification of the concept of *suspendable* in the latest revision supports your position; I've revised the comment.  If the sanctity of cooperative threads gets broken we'll have a problem.

On the other hand internal functions may be written to assume they are
called with interrupts disabled, or a specific lock held.

**TODO** standard marking?

### atomic

An operation is atomic if the steps it makes internally cannot be
affected by nor visible to interleaving executions, such as from
interrupts or thread pre-emption.

An operation that is atomic is by definition interrupt-safe.

An operation that is atomic is by definition thread-safe.

### asynchronous

A function is asynchronous if it may return before the operation it
initiates has completed.  An asynchronous function must provide a
mechanism by which completion is reported, e.g. a callback, event, or
signal.

#### Commentary

Note that asynchronous is orthogonal to non-yielding.  Some API may
provide completion information through a callback, but may yield while
waiting.

### queued

A function is queued if it is asynchronous and allows multiple
operations to be outstanding at any time.

#### Commentary

This concept is necessary because of operations like
`spi_transceive_async()` which returns its result through a signal, but
will yield if the device is already processing an asynchronous
operation.

> Queued operations are rare(? do not exist?) in Zephyr, and require that

Reviewer: No objections to the definition of the term, but I guess it goes without saying that each term should have at least one user before being included in the glossary.

Author: Agreed.  Marked it TBD and proposed, expanded on the reasoning behind why this is relevant, and suggested it's probably the wrong thing to do.  I expect to remove it before this gets merged.

the API use a chainable persisted state object to hold the operation
parameters in a persisted state object that can be added to an internal
queue for processing when the required resource is available.

## Other Rules

> * A function that cannot fulfill its contract when invoked from

Reviewer: Are you sure the overhead is worth it for this one? I'm not convinced it is -- if we document that it's not OK to call something from ISR context, the results of doing so should be UB, I think.

Author: Obviously this needs to be clarified.  I mean that if the function is documented to be callable from an ISR (as, for example, `gpio_pin_write()` should be), then when that generic interface is invoked from an ISR on a particular implementation (an I2C-based GPIO peripheral) that doesn't allow it, the behavior of the API must not only be defined, but be exactly this behavior.  I.e. it is callable, but it may not succeed.

A function that is documented to not be OK to call from an ISR can produce undefined behavior if that restriction is ignored.

Author: This has been addressed by introduction of and commentary on *no-wait*.

A function may succeed or fail.  In many cases failure is indicated by a
negative integer return value while success is indicated by a
non-negative integer return value, but specific functions may use other
conventions.

An **operation** is behavior, which is generally initiated by invoking a
function.  An operation can also succeed or fail.  An operation
**completes** when the success or failure of the operation has been
determined and made available through whatever mechanism was specified
through the initiating function.

For APIs we are particularly interested in the relationship between the
function and its associated operation.  Some potential relationships
include:
* The function succeeds if and only if its associated operation
  completes with success. (This is usually expected of *synchronous*
  functions.)
* If the function fails then its associated operation has not been
  initiated.  If the function succeeds then its associated operation has
  been initiated.  The success or failure of the operation is not
  otherwise coupled to success or failure of its initiating function.
  (This is usually expected of *asynchronous* functions.)

## Definitions

> ### suspendable (function)

Reviewer: We're having trouble finding a word for this that we both like :).

My concern with this latest attempt is that a function which causes the thread to enter the waiting state is suspendable by this definition. This could be confusing, since the reader might thing "suspendable" means "may enter the suspended state".

What about a straightforward "may switch context"?

Author: See new update that uses *rescheduling* and refined terms.

A function is suspendable when invocation by a cooperative thread may
trigger a context switch.

#### Commentary

For thread safety it is generally necessary to hold a lock of some sort.
If a sequence of code involves only invocation of non-suspendable
functions then the lock may be avoidable.

See also discussion of the term "block" at the definition of
*synchronous*.

Note that if a suspendable function is marked as isr-callable then it
must internally check `k_is_in_isr()` and behave as a no-wait function,
whether or not it was invoked in a no-wait mode.

### no-wait (function)

A suspendable function is no-wait when calls to it can be forced by a
parameter into a no-wait mode, i.e. one that takes a path that will not
trigger a context switch from a cooperative thread.

#### Commentary

A no-wait function invoked in no-wait mode in conditions where it cannot
initiate its operation should fail, in many cases by returning the value
`-EWOULDBLOCK`.

> ### isr-callable (function)

Reviewer: I wonder if you want to generalize this notion to encompass being callable from what the Linux kernel calls "atomic context".

Focusing on ISRs might be a bit short-sighted -- what if we're not in an ISR, but do hold a spinlock in thread context on SMP? Would we need "spinlock-callable" too?

Author: That was the thought behind the original question about whether we needed a `k_is_atomic()`.  For now we have only `k_is_in_isr()` and I want two things: (1), a clear definition of what something must promise in order to be unconditionally callable from an ISR (in the latest version that's *reschedule-safe*), and (2) a way to denote that a function allows a parameter to ensure it's callable even when its behavior would normally not be safe (that'd be *no-wait*).

Reviewer: > That was the thought behind the original question about whether we needed a `k_is_atomic()`.

I don't follow that logic, sorry.

The relevant point in the LWN article I keep linking to is that functions can be safe (or not) to call from atomic context, entirely independently of whether or not there exists a predicate that returns true if and only if the current code is running in atomic context.

You seem to be talking about the latter (existence of a predicate). I am talking about the former (whether a code is OK to call from atomic context as a generalization of interrupt context).

What am I missing?

Author: The short version is that I'm focusing on terms that I think I understand.  I don't know what to do about atomic.

Can you explain what you mean by "atomic context"?  If the cases we're trying to address are:
* can be safely invoked from an interrupt handler; versus
* can be safely invoked when a spinlock is held

is there really a relationship between them?

Maybe the thing to do is to start with defining the conditions for something being "atomic-callable".  If you agree, please propose something: I have no idea what that would mean that isn't covered by "reschedule-safe".

Reviewer: Resolved offline, I was seeing things that weren't in the document. Sorry for noise.

  sections of the [Architecture Porting Guide](https://docs.zephyrproject.org/latest/guides/porting/arch.html)

Zephyr defines [six thread
states](https://docs.zephyrproject.org/latest/reference/kernel/threads/index.html#thread-states)
of which four are active:

* **Ready** when there is nothing that prevents the thread from becoming
  active as the current thread, but it is not the current thread.
* **Running** when the thread is active as the current thread (on a
  processor).
* **Waiting** when the thread is on a queue waiting for an event to occur
  that will transition it to *Ready*.
* **Suspended** when the thread is inactive and must be transitioned to
  *Ready* explicitly (e.g. via `k_thread_resume()`.

A thread is made **unready** if it transitions to *Suspended*,
*Waiting*, or *Terminated*.

Zephyr defines [two mechanisms for selecting the running
thread](https://docs.zephyrproject.org/latest/reference/kernel/scheduling/index.html#scheduling):
* In **cooperative** scheduling a thread transitions from *Running* only
  when it explicitly invokes an operation that makes it unready

>   (i.e. *Waiting* or *Suspended*) or invokes `k_yield()`.

Reviewer: ```suggestion
  (i.e. *Suspended*, Waiting*, or *Terminated*), or invokes `k_yield()`.
```

Reviewer: Also, actually, is this true?

What if I call `k_thread_abort(some_coop_thread_id)` from another CPU on SMP?

Not super critical to the terminology below, but it seems like that ought to make the aborted thread unready by these definitions.

Author: Reworded to acknowledge external changes to thread state.

conventions.

An **operation** is behavior, which is generally initiated by invoking a
function.  An operation can also succeed or fail.  An operation
**completes** when the success or failure of the operation has been
determined and made available through whatever mechanism was specified
through the initiating function.

For APIs we are particularly interested in the relationship between the
function and its associated operation.  Some potential relationships
include:
* The function succeeds if and only if its associated operation
  completes with success. (This is usually expected of *synchronous*
  functions.)
* If the function fails then its associated operation has not been
  initiated.  If the function succeeds then its associated operation has
  been initiated.  The success or failure of the operation is not
  otherwise coupled to success or failure of its initiating function.
  (This is usually expected of *asynchronous* functions.)

## Definitions

### rescheduling (function)

A function is rescheduling when it invokes an operation that includes a
reschedule point (potentially, but not necessarily, context switching to
a non-current thread).

> Note that whether a reschedule-point function will cause a context

Reviewer: ```suggestion
Note that whether a rescheduling function will cause a context
```

Author: done

switch depends on the priority of the current and ready threads.

#### Commentary

The term *rescheduling* applies only to functions that explicitly invoke
operations that (may) encounter a reschedule point.  Unless interrupts
are disabled any function may be context-switched if the interrupt
causes a meta-irq thread to become ready.

> ### reschedule-safe (function)

Reviewer: Did you consider calling this "cooperative-safe", and if so, why did you decide against it?

Author: I did not consider that, but it seems to work especially when paired with cooperative-unsafe.

A function is reschedule-safe if it is either not rescheduling, or if it
is rescheduling but no reschedule point reached during its execution
would cause a cooperative thread invoking it to be context-switched.

#### Commentary

For thread safety it is generally necessary to hold a lock of some sort.
If a sequence of code involves only invocation of functions that do not
context-switch then the lock may be avoidable.

See also discussion of the term "block" at the definition of
*synchronous*.

Note that if a function is marked as isr-callable then it
must internally check `k_is_in_isr()` and behave as a no-wait function,
whether or not it was invoked in a no-wait mode.

### context-switching (function)

A function is context-switching when it is rescheduling and is
guaranteed to cause a context switch when the head of the ready queue is
not the current thread, even for cooperative threads.

#### Commentary

Examples include `k_yield()` and `k_sleep()` (TBD: only when passed a
non-zero timeout?).

### no-wait (function)

A rescheduling function that is not reschedule-safe is no-wait when
calls to it can be forced by a parameter into a reschedule-safe path,

> i.e. one that will not trigger a context switch from a cooperative

Reviewer: For example, using "cooperative-safe" would probably make this clause unnecessary.

The above would read "a rescheduling function that is not cooperative-safe is no-wait when calls to it can be forced by a parameter into a cooperative-safe path," which I think is pretty clear on its own.

Author: Revised to use `cooperative-unsafe`.

timeout value `K_NO_WAIT` indicates that the operation should fail
rather than change the thread state.  The failure is indicated with an
operation-specific error code such as `-EAGAIN`, `-EBUSY`, `-ENOMEM`,
and `-ENOMSG`.

* `z_pend_curr` transitions the current thread to **Waiting**;
* `k_yield` transitions the current thread to **Ready** and moves it to
  the end of the ready queue.  It also explicitly overrides the default
  preservation of cooperative threads as the current thread.
* `k_sleep` invokes `k_yield()` if timeout is zero, otherwise suspends
  the thread until the timeout.

An operation cannot be invoked from an ISR if:
* It transitions the current thread to *Suspended* (per assert in
  `z_tick_sleep()`)
* It transitions the current thread to *Waiting* (per documentation of
  e.g. `k_sem_take()`)
* It transitions the current thread to *Ready* (per assert in
  `k_yield()`)

Generalizing this, and consistent with the informal answer to issue
21341, an operation cannot be invoked when invoked from a non-thread
context (interrupt or pre-kernel) if it could cause its invoking thread
to sleep, because without in invoking thread it cannot implement the
required behavior.

> Kernel API assessed using terminology below, plus "cooperative-safe"

Reviewer: This table was very helpful. I would not call it a digression or random note, but maybe promote it up higher.

# Draft material to review in support of [#18970][]

[//]: # (cmark-gfm -e table -e footnotes doc/api18970.md > /tmp/api18970.html)

[#18970]: https://github.com/zephyrproject-rtos/zephyr/issues/18970

Note: See the Background section below for technical concepts necessary
to understand the definitions.

## API Attributes

This work is leading to a proposal for five specific attributes for API
behavior that should be documented in function descriptions and, in some
cases, exposed through API at runtime.

* **reschedule** if executing the function reaches a reschedule point
* **sleep** if executing the function can cause the invoking thread to
  sleep
* **no-wait** if a parameter to the function can prevent the invoking
  thread from trying to sleep

> * **isr-ok** if the function can be safely called from an ISR even if it

Reviewer: ```suggestion
* **isr-ok** if the function can always be safely called from an ISR even if it
```

  may return an error in that case
* **supervisor** if the calling thread must have supervisor privileges
  to execute the function

Details on the behavioral impact of each attribute are in the following
sections.  Attributes for some common kernel API are below:

name              | reschedule | sleep | no-wait | isr-ok | supervisor
----------------- | ---------- | ----- | ------- | ------ | ----------

> `k_busy_wait`     | no         | -     | -       | -      | no

Author: `-` means it's not specified.  If something is not `reschedule`, it's not `sleep`.  If it's not `sleep`, it's not `no-wait`.  That's a consequence of the way the terms are defined. Anything that isn't `reschedule` is assumed to be callable from thread or interrupt context, with caveats.

At this stage I'm focused on the minimum number of attributes necessary to describe behavior in the case where a specific attribute is present.  We can't assume that if `driver_api_foo()` is *not* marked `reschedule` that it can't be pre-empted due to making another thread ready: maybe the way it has to be implemented requires that freedom.  Absence of attribute is not absence of behavior associated with an attribute.

Also see below on why it's not `isr-ok`.

> `k_busy_wait`     | no         | -     | -       | -      | no

Reviewer: > If something is not `reschedule`, it's not `sleep`. If it's not `sleep`, it's not `no-wait`.

Gotcha. Isn't the proper value `no` then for `sleep` and `no-wait`, rather than `-`? From your comment, these *are* specified: the short answer is "no", the long answer is "no by definition, given the value of the `reschedule` column".

> Also see below on why it's not isr-ok

Same here. The answer seems like it's "no" rather than "-".

I would personally have found this clearer with "no" used, at least.

> `k_busy_wait`     | no         | -     | -       | -      | no

Author: I feel the distinction between "attribute is not supported because of the way the function is implemented" (and therefore the behavior could be changed to add the attribute) and "attribute is not supported because it's not relevant" is important.  People who don't understand this will have an incomplete understanding of the relationship between the attributes.

So I've marked these `no*` with `*` documenting that point.

I've also updated `isr-ok` and added a couple more attributes.  Please mark resolved any comments you've made that are no longer relevant.

> `k_busy_wait`     | no         | -     | -       | -      | no

Reviewer: > So I've marked these `no*` with `*` documenting that point.
> 
> I've also updated `isr-ok` and added a couple more attributes.

:+1: 

>  Please mark resolved any comments you've made that are no longer relevant.

Done. The "resolve conversation" button is not available on some of my outdated comments. Please disregard all of those as of time of writing.

Functions that are **sleep** may be invoked from thread context.

Functions that are **sleep** may be invoked from interrupt context if
and only if invoked in **no-wait** mode.

#### Commentary

This attribute is of relevance specifically when considering application
that use only cooperative threads, because the kernel will not replace a
running cooperative thread at a reschedule point unless that thread has
explicitly invoked an operation that caused it to sleep.

Note that this attribute does not imply the function will sleep
unconditionally, but that the state of the system may require the
invoking thread to suspend, wait, or invoke `k_yield()` before it can
complete its operation.  This behavior may be mediated by **no-wait**.

### no-wait

The no-wait attribute is used on a function to indicate that a parameter
to the function can force an execution path that will not cause the
invoking thread to sleep.

Functions with this attribute may be invoked from interrupt context when
the parameter selects the no-wait path.

Functions that are **no-wait** functions are implicitly **sleep**.

### isr-ok

The isr-ok attribute is used on a function to indicate that it can be

> called from interrupt context to the function can force an execution

Reviewer: Parse error; can you please reword?

Author: Something like "can be called from interrupt context ~to the function can~ *and the function will* force an execution".  Basically it's a callee-implemented no-wait.  See details below.

Reviewer: OK... then I think maybe a more complete definition is:

```suggestion
called from interrupt context and function will force an execution
path that will not cause the invoking thread to sleep, regardless
of the values of any of its parameters.
```

The distinction between isr-ok and no-wait seems to be that an isr-ok function is definitionally never going to sleep if called by an ISR (though it might error out), while no-wait can be asked politely not to do so.

Did I understand you?

path that will not cause the invoking thread to sleep.

This attribute is intended for **sleep** functions that may be
indirectly invoked from interrupt context with arguments that would
attempt to cause the invoking thread to sleep.

Functions that are **isr-ok** may be invoked from interrupt context, and

> will return an error if they would have tried to sleep.

Reviewer: Are you sure about making this a hard universal rule? Why not allow certain functions to conform to the historical Zephyr style of not checking for this type of error except via asserts, to avoid overhead?

Author: Because that historical style is exactly why we need this sort of thing.

Take `k_sem_take()`.  You can't call it from an interrupt unless you invoke it in no-wait mode.  If you do the kernel will panic--either because of an assert, or because it's wandered into the fog.  It's "OK" to not return an error because the developer will always know whether the call is in the body of an interrupt handler or a function invoked from a thread, and will have rigorously tested all execution paths with assertions enabled and will be justified in their confidence that the application behaves correctly.

Now consider `gpio_pin_write()`.  You should *always* be able to call that from thread and interrupt context, because it's safe to do that with most SOC controllers and making it illegal would be a major pain in driver code where you need to do something like deassert a signal.

But if the underlying controller identified in a sensor's devicetree node is accessed with I2C then when the driver deasserts the signal from an interrupt handler the GPIO implementation absolutely cannot try to take the semaphore that guarantees mutex of the bus.  So `gpio_pin_write()` must be `isr-ok`, and rather than bring the system down must instead return an error to indicate that it didn't do what it was asked to do.  And the code must propagate that error back to its caller, which must check it and complete its responsibilities some other way.

Now this does reveal an ambiguity that's part of why `k_sem_take()` is not `isr-ok`: because if it was, it would have to do a runtime check, which would get people upset (though the coding guidelines are going to require it anyway).  `gpio_pin_write()` is also not `sleep` or even `reschedule`, even though in certain controllers it will have to sleep.

That's why, once all this stuff gets in place and topic-gpio starts moving forward again, the GPIO API will have to be extended so each gpio controller can provide information about which of its operations are `sleep` so applications can take that into account when interacting with the devices.   (See #17827, which is the fundamental need I've had for the last six months that's motivating this whole effort, and #21090 where the transition operations must be marked as to whether they might sleep so the generic interface can safely avoid panics.)

reschedule point within its execution.

#### Commentary

The significance of this attribute is that when a rescheduling function is
invoked by a thread it is possible for that thread to be suspended as a
consequence of a higher-priority thread being made ready.  Whether the
suspension actually occurs depends on the operation associated with the
reschedule point and the relative priorities of the invoking thread and
the head of the ready queue.

Note that in the case of timeslicing, or reschedule points executed from
interrupts, any thread may be suspended in any function.

Functions that are not **reschedule** may be invoked from either thread
or interrupt context.

Functions that are **reschedule** may be invoked from thread context.

Functions that are **reschedule** but not **sleep** may be invoked from
interrupt context.

### sleep

The sleep attribute is used on a function that can cause the invoking
thread to sleep.

Functions that are **sleep** are implicitly **reschedule**.

Functions that are **sleep** may be invoked from thread context.

Functions that are **sleep** may be invoked from interrupt context if
and only if invoked in **no-wait** mode.

#### Commentary

> This attribute is of relevance specifically when considering application

Reviewer: ```suggestion
This attribute is of relevance specifically when considering applications
```

invoking thread to suspend, wait, or invoke `k_yield()` before it can
complete its operation.  This behavior may be mediated by **no-wait**.

### no-wait

The no-wait attribute is used on a function to indicate that a parameter
to the function can force an execution path that will not cause the
invoking thread to sleep.

Functions with this attribute may be invoked from interrupt context when
the parameter selects the no-wait path.

Functions that are **no-wait** functions are implicitly **sleep**.

#### Commentary

The paradigmatic case of a no-wait function is a function that takes a
timeout, to which `K_NO_WAIT` can be passed.  The semantics of this
special timeout value are to execute the function's operation as long as
it can be completed immediately, and to return an error code rather than
sleep if it cannot.

It is use of the no-wait feature that allows functions like
`k_sem_take()` to be invoked from ISRs, since it is not permitted to
sleep in interrupt context.

**NOTE** That a function has a no-wait path does not imply that taking
that path guarantees the function is synchronous.

### isr-ok

The isr-ok attribute is used on a function to indicate that it can be

> called from interrupt context.  The function will use `k_is_in_isr()` to

Reviewer: ```suggestion
called from interrupt context.  If necessary, the function will use `k_is_in_isr()` to
```

It's not necessary for `k_busy_wait()` to do this, for example.

detect its calling context and force an execution path that will not
cause the invoking thread to sleep.

> This attribute is intended for **sleep** functions that may be

Reviewer: ```suggestion
This attribute is useful for **sleep** functions that may be
```

Just a suggestion, since again `k_busy_wait()` is isr-ok but not sleep.

In general I think it's a bit cleaner if the definition of the term `isr-ok` doesn't assume that functions so marked are `sleep`, even if that's the primary use case.

Author: I've taken the previous suggestion to add "If necessary" regarding `k_is_in_isr()`, but for now I'm going to stick with "intended" because that is its purpose.  "intended" shouldn't imply "reserved for use by".

At this time the only guidance I have for what makes something callable from an interrupt is that "it doesn't sleep".  Nobody's explained what "sleep" means, and my attempts to define it rigorously have led me deep into rabbit warrens so I'm just gonna hand-wave it.

People could tag non-sleep functions with `isr-ok`, but I'm very concerned about a slippery slope where people say "Hey, `k_cycle_get()` (or `k_ms_to_ticks_floor32()` or `crc16()`) isn't marked `isr-ok` so I can't use it in an interrupt?" and we end up explicitly marking a huge number of functions that have nothing to do with thread sleep behavior as `isr-ok`.

If we want complete explicit attribute sets for every API function I'd rather see some tooling that synthesizes derived attributes based on the presence/absence of indicator attributes and populates those in the generated documentation.  Anything that isn't marked `sleeps` should be assumed to be callable from an interrupt.  (Whether doing so is a good idea is a different issue.)

Reviewer: > I'm going to stick with "intended"

ACK

> I'm very concerned about a slippery slope where [...] we end up explicitly marking a huge number of functions that have nothing to do with thread sleep behavior as `isr-ok`.
>
> If we want complete explicit attribute sets for every API function I'd rather see some tooling that synthesizes derived attributes based on the presence/absence of indicator attributes and populates those in the generated documentation. 

OK. Perhaps I'll understand this concern better when you have time to show how you plan on using these terms in a non-draft context to mark up existing APIs in RST documentation.

indirectly invoked from interrupt context with arguments that could
attempt to put the invoking thread to sleep, e.g. because the function
is not **no-wait** or the parameters do not select the no-wait path.

Functions that are **isr-ok** may be always be safely invoked from
interrupt context, and will return an error if they were unable to
fulfill their behavior in that context.

### pre-kernel-ok

The pre-kernel-ok attribute is used on a function that is also

> **isr-ok** to indicate that it will take reasonable steps to ensure it

Reviewer: Looks like the term pre-kernel-ok is a new addition since the last time I looked at this.

Are there any situations in which a function might be pre-kernel-ok, but not isr-ok? I get that in practice they're similar (both "atomic context OK"), but I wonder if it would be better to leave a little wiggle room in the definition by not assuming that `pre-kernel-ok` implies `isr-ok`.

Author: Yes; it's there specifically because there's API in work (#21090) that specifically needs to be safe to invoke before the kernel gets started, since one of its use cases is to bring up the system clock.  However, `k_is_in_isr()` is false before the kernel gets started, so API that wants to promote itself as usable in pre-kernel initialization must do an extra check to avoid invoking API that could sleep.  (Nothing can sleep until the kernel gets past pre-kernel initialization and stops running on the interrupt stack.)

So it's not that something might be pre-kernel-ok but not isr-ok, but rather that something might be isr-ok and not pre-kernel-ok because it doesn't do the full check.

I've added "intended" here to parallel the text in **isr-ok**.

I still have no idea what "atomic context" means, and I don't think we should be using that term without a definition, even informally.

Reviewer: > Yes; it's there specifically because there's API in work (#21090)

Yep, I get why you're adding this.

>> wonder if it would be better to leave a little wiggle room in the definition by not assuming that pre-kernel-ok implies isr-ok.
>
> I've added "intended" here to parallel the text in isr-ok.

I take it from this that you are not interested in this wiggle room. Perhaps this relates to your earlier concern about slippery slopes towards excessive markup, which I think I only half understand.

> I still have no idea what "atomic context" means, and I don't think we should be using that term without a definition, even informally.

I'll propose "any execution context in which it is a programming error to invoke an operation that sleeps" as an informal definition, using the terms in this PR.

Author: > I take it from this that you are not interested in this wiggle room.

I understood you to be saying "don't assume that pre-kernel-ok implies isr-ok". I don't see that implication in the text, and in fact it's not true.  The text did suggest pre-kernel-ok would be *used* only for explicitly isr-ok functions because you *cannot* assume that isr-ok implies pre-kernel-ok.  I've now taken that out here and will do so in #21678 so you could have a pre-kernel-ok function that's not isr-ok; hopefully that's more like what you had in mind.

> > I still have no idea what "atomic context" means....
> 
> I'll propose "any execution context in which it is a programming error to invoke an operation that sleeps" as an informal definition, using the terms in this PR.

For "atomic" I like [this definition](https://wiki.osdev.org/Atomic_operation):

> An atomic operation is an operation that will always be executed without any other process being able to read or change state that is read or changed during the operation. It is effectively executed as a single step....

There's nothing in this that prevents an atomic operation from being interrupted or sleeping as long as those interruptions don't affect the state of the operation.

What you propose above is a useful definition that covers both pre-kernel and interrupt contexts in Zephyr, but I don't see what we're gaining by using "atomic" in conjunction with "context".  I do see that the term is [used in Linux](https://lwn.net/Articles/274695/) as you'd pointed out before, but that reference seems to be an argument for not using it to mean "can't sleep".  Maybe it should be "sleep-disallowed context".

Reviewer: > but I don't see what we're gaining by using "atomic" in conjunction with "context". I do see that the term is [used in Linux](https://lwn.net/Articles/274695/) as you'd pointed out before, but that reference seems to be an argument for not using it to mean "can't sleep". Maybe it should be "sleep-disallowed context".

I'm happy to stop borrowing the term "atomic context" from Linux -- which I'm only using informally and am not proposing we include in this list of terms -- if you are unhappy with it. I just use it because it's familiar shorthand from Linux, and Zephyr borrows a lot from there.

I would like to have an informal term we can use for "times when you can't sleep," since I think it's a generally useful notion, as shown by the inclusion of pre-kernel-ok and isr-ok in this list of terms.

With apologies to the Beastie Boys, we can call it "on our way to Brooklyn" mode if you want; I don't care :).

Reviewer: > I understood you to be saying "don't assume that pre-kernel-ok implies isr-ok". I don't see that implication in the text, and in fact it's not true. The text did suggest pre-kernel-ok would be _used_ only for explicitly isr-ok functions because you _cannot_ assume that isr-ok implies pre-kernel-ok. I've now taken that out here and will do so in #21678 so you could have a pre-kernel-ok function that's not isr-ok; hopefully that's more like what you had in mind.

Yep, that's clearer to me now, thanks.

Author: Can we just call it "non-thread context" and define it as being the union of pre-kernel and interrupt context, rather than relying on the concept of "can't sleep" for its definition?  Those two plus thread context are, AFAIK, the only contexts available.

Note that in the final version I don't try to specify what "sleep" means.  That's because it becomes something like "an operation sleeps when, if it were invoked in thread context, it would cause the running thread to transition itself to a non-running state" (which could be suspended, waiting, ready-but-not-head-of-ready-queue).  Which is horrific, and probably subtly inaccurate.

Reviewer: > Can we just call it "non-thread context" and define it as being the union of pre-kernel and interrupt context, rather than relying on the concept of "can't sleep" for its definition? Those two plus thread context are, AFAIK, the only contexts available.

I'm fine with using the term "non-thread context" between ourselves, but that's not exactly what I'm referring to when I say "atomic context" -- the latter applies even in thread context whenever a spinlock is held.

Author: Ah.  In that case I'd want "spin-locked context" as a more useful primitive concept, and you can define "atomic context" as "non-thread or spin-locked".  Though the way I undersatnd "atomic" corresponds more to "spin-locked" alone: interrupts aren't inherently atomic since they can be nested.

AFAIK, you can't sleep in an ISR because it just won't work to sleep while running on the interrupt stack.  You can sleep while holding a spin lock, but there's a chance you'll deadlock.  (ZLIs and possibly meta-IRQ threads becoming enabled break IRQ locks, so you might be able to do it in some perverse but functional way even without SMP).

Reviewer: > AFAIK, you can't sleep in an ISR because it just won't work to sleep while running on the interrupt stack. You can sleep while holding a spin lock, but there's a chance you'll deadlock.

Right, which is why I used "it is a programming error" in my definition, which I believe applies to both cases.

>  In that case I'd want "spin-locked context" as a more useful primitive concept, and you can define "atomic context" as "non-thread or spin-locked".

As far as I know, that's a fine definition for Zephyr on Arm Cortex-M. :thumbsup:
